internal/core/cluster.go

```diff
@@ -85,86 +93,60 @@ func (c *Cluster) Connect(ctx context.Context, provider *gophercloud.ProviderCli
-		err := c.LiquidConnections[serviceType].reconcileLiquidConnection()
+		conn := c.LiquidConnections[serviceType]
+		err := conn.Init(ctx, provider, eo)
 		if err != nil {
-			return err
+			errs.Addf("failed to initialize service %s: %w", serviceType, util.UnpackError(err))
 		}
 	}
 
 	// delete all orphaned cluster_services
 	// respective cluster_resources, cluster_az_resources and cluster_rates are handled by delete-cascade
 	_, err = c.DB.Exec(`DELETE FROM cluster_services WHERE type != ALL($1)`, pq.Array(serviceTypes))
 	if err != nil {
-		return fmt.Errorf("cannot cleanup orphaned cluster_services: %w", err)
+		errs.Addf("failed orphaned cluster_services cleanup: %w", err)
 	}
-	return nil
+
+	return errs
 }
 
-// ServiceTypesInAlphabeticalOrder can be used when service types need to be
-// iterated over in a stable order (mostly to ensure deterministic behavior in unit tests).
-func (c *Cluster) ServiceTypesInAlphabeticalOrder() []db.ServiceType {
-	result := make([]db.ServiceType, 0, len(c.LiquidConnections))
+// AllServiceInfos returns a map of all ServiceInfos for all services in this cluster.
+// Its output is the basis to use the convenience methods below to get certain properties of the services
+// configuration. In order to be usable efficiently, it is recommended to call this method only once per API,
+// so that the database fallback is only done once.
+func (c *Cluster) AllServiceInfos() (map[db.ServiceType]liquid.ServiceInfo, error) {
+	if len(c.LiquidConnections) == 0 {
+		return readServiceInfoFromDB(c.DB, None[db.ServiceType]())
+	}
+	result := make(map[db.ServiceType]liquid.ServiceInfo, len(c.LiquidConnections))
 	for serviceType, connection := range c.LiquidConnections {
 		if connection != nil { // defense in depth (nil values should never be stored in the map anyway)
-			result = append(result, serviceType)
+			result[serviceType] = connection.ServiceInfo()
 		}
 	}
-	slices.Sort(result)
-	return result
-}
-
-// HasService checks whether the given service is enabled in this cluster.
-func (c *Cluster) HasService(serviceType db.ServiceType) bool {
-	return c.LiquidConnections[serviceType] != nil
+	return result, nil
 }
 
-// HasResource checks whether the given service is enabled in this cluster and
-// whether it advertises the given resource.
-func (c *Cluster) HasResource(serviceType db.ServiceType, resourceName liquid.ResourceName) bool {
-	connection := c.LiquidConnections[serviceType]
-	if connection == nil {
-		return false
+// InfoForService returns the ServiceInfo for a given service. If the service does not
+// exist, an empty map is returned. It should be used instead of Cluster.AllServiceInfos
+// when only one service is needed, to avoid the overhead of loading all services. It has the same
+// signature as Cluster.AllServiceInfos to allow to use the utility types, but only returns a single entry
+func (c *Cluster) InfoForService(serviceType db.ServiceType) (map[db.ServiceType]liquid.ServiceInfo, error) {
```

Comment:
This is a weird return type. Since we can either have one or zero ServiceInfo results, I suggest:
```suggestion
func (c *Cluster) InfoForService(serviceType db.ServiceType) (Option[liquid.ServiceInfo], error) {
```